File: qt/gen_qsyntaxhighlighter.go
package qt

import (
	"unsafe"
)

type QSyntaxHighlighter struct {
	h          uintptr
	isSubclass bool
}

// NewQSyntaxHighlighter constructs a new QSyntaxHighlighter object.
func NewQSyntaxHighlighter(parent *QObject) *QSyntaxHighlighter {
	ret := newQSyntaxHighlighter(QSyntaxHighlighter_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQSyntaxHighlighter2 constructs a new QSyntaxHighlighter object.
func NewQSyntaxHighlighter2(parent *QTextDocument) *QSyntaxHighlighter {
	ret := newQSyntaxHighlighter(QSyntaxHighlighter_new2(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QSyntaxHighlighter) MetaObject() *QMetaObject {
	return newQMetaObject(QSyntaxHighlighter_MetaObject(this.h))
}

func (this *QSyntaxHighlighter) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QSyntaxHighlighter_Metacast(this.h, param1_Cstring))
}

func QSyntaxHighlighter_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QSyntaxHighlighter_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QSyntaxHighlighter) SetDocument(doc *QTextDocument) {
	QSyntaxHighlighter_SetDocument(this.h, doc.cPointer())
}

func (this *QSyntaxHighlighter) Document() *QTextDocument {
	return newQTextDocument(QSyntaxHighlighter_Document(this.h))
}

func (this *QSyntaxHighlighter) Rehighlight() {
	QSyntaxHighlighter_Rehighlight(this.h)
}

func (this *QSyntaxHighlighter) RehighlightBlock(block *QTextBlock) {
	QSyntaxHighlighter_RehighlightBlock(this.h, block.cPointer())
}

func QSyntaxHighlighter_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QSyntaxHighlighter_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QSyntaxHighlighter_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QSyntaxHighlighter_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QSyntaxHighlighter) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QSyntaxHighlighter_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QSyntaxHighlighter) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QSyntaxHighlighter_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSyntaxHighlighter_MetaObject
func miqt_exec_callback_QSyntaxHighlighter_MetaObject(self QSyntaxHighlighter, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QSyntaxHighlighter{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QSyntaxHighlighter) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QSyntaxHighlighter_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QSyntaxHighlighter) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QSyntaxHighlighter_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSyntaxHighlighter_Metacast
func miqt_exec_callback_QSyntaxHighlighter_Metacast(self QSyntaxHighlighter, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QSyntaxHighlighter{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qsyntaxhighlighter.cpp
// +build ignore

#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QSyntaxHighlighter>
#include <QTextBlock>
#include <QTextDocument>
#include <qsyntaxhighlighter.h>
#include "gen_qsyntaxhighlighter.h"
class MiqtVirtualQSyntaxHighlighter : public virtual QSyntaxHighlighter {
public:
MiqtVirtualQSyntaxHighlighter(QObject* parent): QSyntaxHighlighter(parent) {};
MiqtVirtualQSyntaxHighlighter(QTextDocument* parent): QSyntaxHighlighter(parent) {};

virtual ~MiqtVirtualQSyntaxHighlighter() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QSyntaxHighlighter::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QSyntaxHighlighter_MetaObject(const_cast<MiqtVirtualQSyntaxHighlighter*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QSyntaxHighlighter::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QSyntaxHighlighter::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QSyntaxHighlighter_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QSyntaxHighlighter::qt_metacast(param1);

	}
};
QSyntaxHighlighter* QSyntaxHighlighter_new(QObject* parent) {
return new MiqtVirtualQSyntaxHighlighter(parent);
}
QSyntaxHighlighter* QSyntaxHighlighter_new2(QTextDocument* parent) {
return new MiqtVirtualQSyntaxHighlighter(parent);
}
void QSyntaxHighlighter_virtbase(QSyntaxHighlighter* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QSyntaxHighlighter_MetaObject(const QSyntaxHighlighter* self) {
	return (QMetaObject*) self->metaObject();
}
void* QSyntaxHighlighter_Metacast(QSyntaxHighlighter* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QSyntaxHighlighter_Tr(const char* s) {
	QString _ret = QSyntaxHighlighter::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QSyntaxHighlighter_SetDocument(QSyntaxHighlighter* self, QTextDocument* doc) {
	self->setDocument(doc);
}
QTextDocument* QSyntaxHighlighter_Document(const QSyntaxHighlighter* self) {
	return self->document();
}
void QSyntaxHighlighter_Rehighlight(QSyntaxHighlighter* self) {
	self->rehighlight();
}
void QSyntaxHighlighter_RehighlightBlock(QSyntaxHighlighter* self, QTextBlock* block) {
	self->rehighlightBlock(*block);
}
struct miqt_string QSyntaxHighlighter_Tr2(const char* s, const char* c) {
	QString _ret = QSyntaxHighlighter::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QSyntaxHighlighter_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSyntaxHighlighter::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QSyntaxHighlighter_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQSyntaxHighlighter*>( (QSyntaxHighlighter*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QSyntaxHighlighter_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQSyntaxHighlighter*)(self) )->virtualbase_MetaObject();
}
void QSyntaxHighlighter_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQSyntaxHighlighter*>( (QSyntaxHighlighter*)(self) )->handle__Metacast = slot;
}
void* QSyntaxHighlighter_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQSyntaxHighlighter*)(self) )->virtualbase_Metacast(param1);
}
void QSyntaxHighlighter_Delete(QSyntaxHighlighter* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQSyntaxHighlighter*>( self );
	} else {
		delete self;
	}
}
